### PR TITLE
Add

### PR DESCRIPTION
**Pull request by kosmos30732**

### CODE REVIEW

File: bb.c [hunk 1]
```diff
@@ -0,0 +1,418 @@
...
+ 
+    //checking if the offset is less than one
+    if (size_of_shift <1)
+    {
+        bb_errno = ERR_SMALL_SIZE;
+        return ERR_SMALL_SIZE;
+    }  
+ 
+    //shift
+    for (int j = 0; j < size_of_shift; j++)
```
Comment (yalegko) [inline bb.c]: Сдвигать большой вектор сдвигами по 1 это как-то уж слишком неэффективно.

Классический вариант: один цикл, который перекладывает базы+делает в них сдвиги за один пробег, либо два цикла, один сдвигает базы, второй делает сдвиги внутри баз

File: bb.c [hunk 2]
```diff
@@ -0,0 +1,418 @@
...
+
+ int bb_errno=ERR_OK;
+
+int left_shift (bb* vector, int size_of_shift)
+{
+    //checking if the shift is possible
+    if (size_of_shift >(legth_of_parts-1-vector->last_part)*8+7-vector->last_bit)
+    {
+        bb_errno = ERR_BIG_SHIFT;
+        return ERR_BIG_SHIFT;
```
Comment (yalegko) [inline bb.c]: errno переменные удобно использовать, когда не хочется усложнять сигнатуру функции отдельным аргументом для результата.

Например, `fopen` возвращает FILE* и потому сообщает об ошибке через NULL в результате + errno. Конструкция `status = fopen("file.txt", "r", &file)` была бы менее выразительна при использовании (но тем не менее есть огромное количество бибилотек, которые всегда возвращают статус, а результаты отдают по указателям)

То есть если есть возможность вернуть статус, то явно указываются (в комментариях) статусы, которые может вернуть функция и что они означают, а errno не выставляется.

Если вариантов вернуть статус нет (не хочется усложнять сигнатуру функции), то пишется что функция вернёт в случае ошибки (обязательно!), e.g. NULL у `fopen` и какие статусы errno она выствляет и в каком случае

File: bb.c [hunk 3]
```diff
@@ -0,0 +1,418 @@
...
+    int need_shift = 0;
+ 
+    //index for the string
+    int ind_string = 0;
+ 
+    //reading bits from a string and writing them to a vector
+    while (string[ind_string]!='\0')
+    {
+        //checks if the string contains something other than 1 and 0
+        if (string[ind_string] == '1' || string[ind_string] == '0')
```
Comment (yalegko) [inline bb.c]: Всегда имеет смысл проверять более короткую ветку условий: код от этого становится сильно читаемее, так как уходит лишняя вложенность

File: bb.c [hunk 4]
```diff
@@ -0,0 +1,418 @@
...
+            if (string[ind_string] == '1')
+            {
+                vector->parts[0] |= (1 << 0);
+            }
+            //don't need to add a zero because when you shift it is added
+        }
+        //if it contains it we output a message
+        else
+        {
+            return NULL;
```
Comment (yalegko) [inline bb.c]: Вот это как раз пример функции, где стоит использовать `errno` (либо возвращать результат по указателю и сделать статусы)

Пользователь должен явно понимать что не так с его входными данными.

File: bb.c [hunk 5]
```diff
@@ -0,0 +1,418 @@
...
+                {
+                    return NULL;
+                }
+            }
+            need_shift = 1;
+ 
+            //adding a bit to the end
+            if (string[ind_string] == '1')
+            {
+                vector->parts[0] |= (1 << 0);
```
Comment (yalegko) [inline bb.c]: А почему 1<<0 а не 1?)

File: bb.c [hunk 6]
```diff
@@ -0,0 +1,418 @@
...
+        if (string[ind_string] == '1' || string[ind_string] == '0')
+        {
+            if (need_shift)
+            {
+                if (left_shift (vector, 1))
+                {
+                    return NULL;
+                }
+            }
+            need_shift = 1;
```
Comment (yalegko) [inline bb.c]: Наверное можно попробовать обработать первый бит отдельно и тогда цикл станет аккуратнее 🤔

File: bb.c [hunk 7]
```diff
@@ -0,0 +1,418 @@
...
+    }
+
+    return vector;
+}
+ 
+char* bb_to_string (bb* vector)
+{
+    //initializing a string and its index
+    int ind_string = 0;
+    char* string = malloc ((ind_string + 1) * sizeof (char));
```
Comment (yalegko) [inline bb.c]: По-хорошему надо проверить что память выделилась, иначе можно неприятно упасть на обращении по NULL

File: bb.c [hunk 8]
```diff
@@ -0,0 +1,418 @@
...
+    {
+        //checking which bit is 1 or 0
+        if (vector->parts[vector->last_part]&(1<<i))
+        {
+            string[ind_string] = '1';
+        }
+        else
+        {
+            string[ind_string] = '0';
+        }
```
Comment (yalegko) [inline bb.c]: Можно немного побыть хардовым сишником и сделать 

```c
char bit = vector->parts[vector->last_part]&(1<<i);
string[ind_string] = '0' + bit;
```

:)

File: bb.c [hunk 9]
```diff
@@ -0,0 +1,418 @@
...
+ 
+        //after writing the bit, re-create the string by +1 element
+        ind_string++;
+        string = (char*)realloc (string, (ind_string + 1) * sizeof (char));
```
Comment (yalegko) [inline bb.c]: Совершенно лишнее, ты ведь и так знаешь сколько тебе символов понадобится

File: bb.c [hunk 10]
```diff
@@ -0,0 +1,418 @@
...
+        ind_string++;
+    }
+
+    return vector;
+}
+ 
+char* bb_to_string (bb* vector)
+{
+    //initializing a string and its index
+    int ind_string = 0;
```
Comment (yalegko) [inline bb.c]: Стандартное сокращение для индекса вне цикла -- `idx`.

Ещё в английском важен порядок слов, поэтому согласование `признак + предмет` надо писать именно в таком порядке, т.е. `string_idx`

File: bb.c [hunk 11]
```diff
@@ -0,0 +1,418 @@
...
+    return string;
+}
+ 
+bb* bb_disjunction (bb* vector1, bb* vector2)
+{
+    //Checking that the vectors are the same length
+    if (vector1->last_part!=vector2->last_part || vector1->last_bit!=vector2->last_bit)
+    {
+        return NULL;
+    }
```
Comment (yalegko) [inline bb.c]: Предполагается, что мы как и в математике можем представлять "ведущие нули" у коротких векторов и дизъюнкцию можно делать у векторов разного размера

```
10101 | 1010 == 11111
```

File: bb.c [hunk 12]
```diff
@@ -0,0 +1,418 @@
...
+
+    bb* vector = malloc (sizeof (bb));
+    vector->last_bit = vector1->last_bit;
+    vector->last_part = vector1->last_part;
```
Comment (yalegko) [inline bb.c]: Вполне понимаю соблазн вернуть результат операции в return, но вот таким образом нам придётся следить за всеми результатами и делать free на каждый, что практически нереально а цепочках операций, e.g

```c
bb_xor(bb_conjunction(bb_from_string("10101"), bb_from_string("111")), bb_from_string("101010101"))
```

В данной конкретной работе у нас есть ограничение по максимальному размеру вектора, поэтому можно просто вернуть саму структуру со стека и она будет скопирована компилятором (тк буффер статический и полностью содержится в структуре)

File: bb.c [hunk 13]
```diff
@@ -0,0 +1,418 @@
...
+    //checking if the offset is less than one
+    if (size_of_shift < 1)
+    {
+        bb_errno = ERR_SMALL_SIZE;
+        return ERR_SMALL_SIZE;
+    }
+ 
+    //cycle left shift
+    for (int i = 0; i < size_of_shift; i++)
+    {
```
Comment (yalegko) [inline bb.c]: Также: сдвигать по 1 биту так себе затея. Лучше обойтись одним-двумя циклами.

Кроме того можно хитро обойти уже имеющимися операциями и почти не писать код сдвига :)

File: bb.h [hunk 14]
```diff
@@ -0,0 +1,42 @@
...
+{
+    uint8_t parts[legth_of_parts];
+    int last_bit;
+    int last_part;
+};
+ 
+typedef struct BigBool bb;
+
+//shifts
+int left_shift (bb* vector, int size_of_shift); //makes a left shift by size_of_shift and return zero in case of successful execution
```
Comment (yalegko) [inline bb.h]: Комментарии к функциям пишут сверху, часто в multi-line комменте. 

Помимо того, что делает и что принимает, обязательно дописать возвращаемые значения, особенно при ошибках. Плюс стоит описать кто владеет возвращаемой памятью. Например, если функция возвращает bb* то надо писать должен ли пользователь эту память очищать и как

Места где можно сразу посмотреть чтобы всё было хорошо я чёт сходу придумать не могу, но вот пара приличных ссылок:
- Достаточно неплохая либа на Си, которой я пользовался не давно. Не супер читаемо, но инфы в комменте хватает: https://github.com/libyal/libregf/blob/master/include/libregf.h.in
- Секция Function comments тут: https://www.cs.swarthmore.edu/~newhall/unixhelp/c_codestyle.html
- Как пример того, что стоит указать при описании функции: https://en.cppreference.com/w/c/io/fopen

**Review by yalegko — CHANGES_REQUESTED** [review]

File: bb.c (hunk 1 above)
Comment (kosmos30732) [reply bb.c]: А подробнее можно? не могу врубиться что за переложение баз... Под базой понимать байт который сдвинется на другой байт

File: bb.c (hunk 3 above)
Comment (kosmos30732) [reply bb.c]: Так я всего лишь проверяю наличие 1 или 0 он и так короткий, как еще его укоротить?

File: bb.c (hunk 5 above)
Comment (kosmos30732) [reply bb.c]: Потому что в parts'ы биты записываются от 0 до 7

File: bb.c (hunk 8 above)
Comment (kosmos30732) [reply bb.c]: Это обязательно именно так?

File: bb.c (hunk 1 above)
Comment (yalegko) [reply bb.c]: Сдвиг базы -- когда i-й элемент двигается на j-ю позицию
Сдвиг внутри базы -- сдвиг битов внутри одного элемента массива.

В рамках текущей лабы можно просто читать "база" как байт, но как я уже говорил на паре, реализация больших чисел эффективнее с большими базами, uint32/uint64

File: bb.c (hunk 3 above)
Comment (yalegko) [reply bb.c]: Не условие, а сами ветки. 

```c
if (!ok) 
{
    return STATUS_NOK;
}
// continue stuff
```

File: bb.c (hunk 5 above)
Comment (yalegko) [reply bb.c]: `1<<0 == 1`

File: bb.h (hunk 14 above)
Comment (kosmos30732) [reply bb.h]: А в самих функциях можно добавить чистку и прыгать goto?

Comment (yalegko) [reply bb.h]: > А в самих функциях можно добавить чистку и прыгать goto?

Если это необходимо, то да. Как я и говорил: необходимость очистки ресурсов при выходе во время ошибки один из нормальных юзкейсов goto

File: bb.c (hunk 11 above)
Comment (kosmos30732) [reply bb.c]: При создании векторы все равно нулями заполняются, тогда есть смысл нули вручную дописывать до максимальной из двух длин векторов? или просто тогда оставить строки 270-273?

File: bb.h [hunk 15]
```diff
@@ -0,0 +1,213 @@
+#ifndef BB_H
+#define BB_H
+
+#include<stdio.h>
+#include<stdlib.h>
+#include<inttypes.h>
+
+#define legth_of_parts 20
```
Comment (yalegko) [inline bb.h]: Дефайны всегда пишет ALL_CAPS чтобы было явно видно что это константа. Иначе можно как я сейчас сидеть понимать что произошло и откуда взялась переменная

File: bb.h [hunk 16]
```diff
@@ -0,0 +1,213 @@
...
+
+#define ERR_OK 0
+#define ERR_SMALL_SHIFT -1
+#define ERR_BIG_SHIFT 1
+#define ERR_MEM_NOT_ALLOC 4
+#define ERR_WRONG_CHARS 2
+#define ERR_NULL_INPUT 3
+
+//bb_errno is set by library functions when error occurs.
+extern int bb_errno;
```
Comment (yalegko) [inline bb.h]: Чтобы было совсем хорошо, можно сделать отдельный тип для статусов и возвращать его из функций.

Это обычно сильно помогает понять где возвращается статус, а где просто число (ну и в целом type-safety -- круто)

File: bb.c [hunk 17]
```diff
@@ -0,0 +1,423 @@
...
+    //reading bits from a string and writing them to a vector
+    while (string[string_idx] != '\0')
+    {
+        //checks if the string contains something other than 1 and 0
+        if (string[string_idx] == '1' || string[string_idx] == '0')
+        {
+			vector=left_shift (vector, 1);
+
+            //adding a bit to the end
+			vector->parts[0] |= (string[string_idx]-48);
```
Comment (yalegko) [inline bb.c]: Если где-то ты делаешь арифметику с кодом символа, то всегда пиши именно символ, а не числовой код. 

`string[string_idx]-'0'`

File: bb.c [hunk 18]
```diff
@@ -0,0 +1,423 @@
...
+    vector->last_bit = 0;
+    vector->last_part = 0;
+
+    //index for the string
+    int string_idx = 1;
+
+	//read first bit
+	if (string[0]!='\0')
+	{
+		if (string[0]=='1')
```
Comment (yalegko) [inline bb.c]: Через `&&` наверное смотрелось бы аккуратнее)

File: bb.c [hunk 19]
```diff
@@ -0,0 +1,423 @@
...
+{
+    if (vector == NULL)
+    {
+        bb_errno = ERR_NULL_INPUT;
+        return NULL;
+    }
+
+    //initializing a string and its index
+    int string_idx = 0;
+    char* string = calloc (1,(vector->last_part*8+vector->last_bit) * sizeof (char));
```
Comment (yalegko) [inline bb.c]: 0-байт кажется не посчитал

File: bb.c (hunk 11 above)
Comment (yalegko) [reply bb.c]: Если ты уверен, что всегда, когда вектор уменьшается, его старые разряды заполняются нулями, то можно явно не описывать этот кейс.

Но это потенциально место для бага

File: bb.c (hunk 12 above)
Comment (yalegko) [reply bb.c]: Вопрос о динамической памяти всё ещё открыт. С аллокацией в каждой процедуре имхо медленно делать операции и сложно следить за тем, чтобы память не текла.

Это не требование, но сразу предупреждаю, что с текущим вариантом в тестах буду пристально следить, чтобы вся память была почищена

File: bb.c [hunk 20]
```diff
@@ -0,0 +1,423 @@
...
+bb* bb_conjunction (bb* vector1, bb* vector2)
+{
+    if (vector1 == NULL || vector2 == NULL)
+    {
+        bb_errno = ERR_NULL_INPUT;
+        return NULL;
+    }
+
+    //if the length of the second vector is longer than the first one then swap them
+    if (vector2->last_part * 8 + vector2->last_bit > vector1->last_part * 8 + vector1->last_bit)
```
Comment (yalegko) [inline bb.c]: Я бы сделал инлайн-функцию для получения длины в битах и пользовался тут ей, было бы весьма читаемо и красиво

File: bb.c [hunk 21]
```diff
@@ -0,0 +1,423 @@
...
+
+    if (vector == NULL)
+    {
+        bb_errno = ERR_NULL_INPUT;
+        return NULL;
+    }
+
+    bb* new_vector1 = left_shift (vector, size_of_shift);
+    bb* new_vector2 = rigth_shift (vector, vector->last_part * 8 + vector->last_bit - size_of_shift);
+    bb* new_vector = bb_disjunction (new_vector1, new_vector2);
```
Comment (yalegko) [inline bb.c]: Не проверяешь ретёрны/статусы операций

**Review by yalegko — CHANGES_REQUESTED** [review]
Не считая пары мелочей код выглядит очень здорово.

Самое время написать тесты и проверить рабочий ли он и удобно ли им пользоваться :)

File: bb.c (hunk 21 above)
Comment (kosmos30732) [reply bb.c]: А как их лучше проверить: После вызова проверить что bb_errno == ERR_OK и идти дальше, а если нет то что.... новую ошибку создать типа ERR_WITH_LEFT_CYCLE? Или вовсе проверять каждую ошибку после каждой функции и если она равна этой возвращать эту же ошибку и NULL?

File: bb.c (hunk 19 above)
Comment (kosmos30732) [reply bb.c]: нет, он включен

File: bb.c [hunk 22]
```diff
@@ -0,0 +1,467 @@
...
+    new_vector->last_bit = vector->last_bit;
+
+    return new_vector;
+
+free3:
+    free (new_vector);
+free2:
+    free (new_vector2);
+free1:
+    free (new_vector1);
```
Comment (yalegko) [inline bb.c]: В виндовс-библиотеках часто встречается вариант

```c
cleanup:
     if (new_vector != NULL) free(new_vector);
     if (new_vector2 != NULL) free(new_vector2);
```

Но на твой вкус

File: bb.c [hunk 23]
```diff
@@ -0,0 +1,467 @@
...
+    }
+
+    if (vector == NULL)
+    {
+        bb_errno = ERR_NULL_INPUT;
+        return NULL;
+    }
+
+    bb* new_vector1 = left_shift (vector, size_of_shift);
+    if (bb_errno!=ERR_OK)
```
Comment (yalegko) [inline bb.c]: Кажется у тебя API говорит проверять errno только в случае нулевого результата

File: tests.c [hunk 24]
```diff
@@ -0,0 +1,118 @@
...
+	bb* vector = calloc (1, sizeof (bb));;
+	vector->last_bit = 1;
+	vector->last_part = 0;
+	vector->parts[0] = 2;
+
+	if (strcmp(bb_to_string (vector),"10")!=0)
+	{
+		LOG_ERR ("bb_to_string not work (10!=10)");
+		return 1;
+	}
```
Comment (yalegko) [inline tests.c]: Приватные штуки типа того как вектор реально лежит в памяти обычно не тестируют. 

Чтобы это проверить стоит сделать конструктор от `uint64_t` и сравнивать два конструктора между собой

File: tests.c [hunk 25]
```diff
@@ -0,0 +1,118 @@
...
+		LOG_ERR ("Don't set ERR_BIG_SHIFT");
+		return 1;
+	}
+
+	rigth_shift (bb_from_string ("11010"), -2);
+	if (bb_errno != ERR_SMALL_SHIFT)
+	{
+		LOG_ERR ("Don't set ERR_SMALL_SHIFT");
+		return 1;
+	}
```
Comment (yalegko) [inline tests.c]: Напрашивается разбиение на отдельные функции (тест-кейсы), проверяющие конкретные блоки кода (конструкторы, сдвиги, ошибки)

File: tests.c [hunk 26]
```diff
@@ -0,0 +1,118 @@
...
+	{
+		LOG_ERR ("cycle_left_shift not work (10010100000 -3> != 10100000100)");
+		return 1;
+	}
+
+	if (strcmp (bb_to_string (cycle_rigth_shift (bb_from_string ("101100001"), 5)), "000011011") != 0)
+	{
+		LOG_ERR ("cycle_rigth_shift not work (101100001 -5> != 000011011)");
+		return 1;
+	}
```
Comment (yalegko) [inline tests.c]: Тесты, как правило, пишутся таким образом, чтобы показать пользователю примеры правильного использования библиотеки.

Сейчас они написаны так, что память, выделенная на куче, никогда не освобождается, что является грубой ошибкой.

Нужно или пересмотреть способ возврата вектора из функций, либо чистить всю выделяемую память

File: tests.c [hunk 27]
```diff
@@ -0,0 +1,118 @@
...
+	{
+		LOG_ERR ("rigth_shift not work (101100001 -5> != 1011)");
+		return 1;
+	}
+
+	if (strcmp (bb_to_string (bb_inversion (bb_from_string ("1011001"))), "0100110") != 0)
+	{
+		LOG_ERR ("bb_inversion not work (1011001-> != 0100110)");
+		return 1;
+	}
```
Comment (yalegko) [inline tests.c]: Напрашивается какой-нибудь макрос/функция типа 

```c 
int bb_eq_str(bb_inversion (bb_from_string ("1011001")), "0100110")
```

**Review by yalegko — CHANGES_REQUESTED** [review]

File: tests.c [hunk 28]
```diff
@@ -0,0 +1,236 @@
...
+	return count;
+}
+
+int tests_transform ()
+{
+	int count = 0;
+	bb *vec1 = NULL, *vec2 = NULL;
+	char* str1 = NULL, * str2 = NULL;
+
+	vec1 = bb_from_uint64_t (1573815);
```
Comment (yalegko) [inline tests.c]: Вроде число можно записать прям в бинарной СС: https://gcc.gnu.org/onlinedocs/gcc/Binary-constants.html

Если нет, то лучше обычно предпочесть 16-чную, у читателя появится небольшой шанс перевести его в биты "на глаз"

File: tests.c [hunk 29]
```diff
@@ -0,0 +1,236 @@
...
+
+	if (vec1 != NULL) free (vec1);
+	if (vec2 != NULL) free (vec2);
+
+	return count;
+}
+
+int tests_shifts ()
+{
+	bb* vec = bb_from_string ("100001010100110101");
```
Comment (yalegko) [inline tests.c:74]: Хорошо бы ещё проверить corner cases, например, что вектор можно "сдвинуть направо" до 0.

Сдвинуть налево целиком во все 0 (типа 1 << 100) в твоей реализации не получится, да?

File: tests.c [hunk 30]
```diff
@@ -0,0 +1,236 @@
...
+	if (vec1 != NULL) free (vec1);
+	if (vec2 != NULL) free (vec2);
+	if (str1 != NULL) free (str1);
+	if (str2 != NULL) free (str2);
+
+
+	return count;
+}
+
+int main ()
```
Comment (yalegko) [inline tests.c:322]: Помимо чистых юнит-тестов имеет смысл сделать тесты, которые генерируют случайные входные данные и потом проверяют какие-то булевы тождества, например, `(x & ~x) == 0`, сдвиги сначала влево, а потом вправо тоже приводят к оригинальному результату, можно выразить XOR через булев базис и проверять его

**Review by yalegko — CHANGES_REQUESTED** [review]

File: tests.c [hunk 31]
```diff
@@ -0,0 +1,346 @@
...
+	}
+	if (output_string != NULL) free (output_string);
+	if (output_vec != NULL) free (output_vec);
+
+	if (vec != NULL) free (vec);
+
+	return count;
+}
+
+int tests_with_random ()
```
Comment (yalegko) [inline tests.c:252]: В несколько отдельных функций в комментариями какое именно тождество проверяется было бы читаемее

**Review by yalegko — APPROVED** [review]
Отлично, зачтено
